pkg/controllers/provisioningv2/rke2/bootstrap/controller.go

```diff
@@ -105,9 +105,14 @@ func (h *handler) getBootstrapSecret(namespace, name string, envVars []corev1.En
 
 	if err != nil {
 		return nil, err
-
 	}
+
 	for _, secretRef := range sa.Secrets {
+		machine, err := h.machineCache.Get(namespace, name)
+		if machine == nil || err != nil {
+			return nil, err
+		}
```

Comment:
This is not dependent on the loop. Move it outside the loop so it only happens once (if there are multiple secret references).